### PR TITLE
Created pushHeadless sample app for push registration via headless on INSTALL vs on first launch.

### PR DESCRIPTION
**Pull request by jamesjhedley**

### CODE REVIEW

**Comment by kwallis** [comment]
Does it only run on install? It would need to run every time the device boots as well would it not?

**Comment by jamesjhedley** [comment]
Well technically the registration should only ever need to run once (on install) but yea the headless app should also run after each device boot :).

File: pushHeadless/README.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
+# Push Headless Sample Application
+
+The Push Headless sample push-enabled application demonstrates how to write a BlackBerry 10 WebWorks application that uses a backend headless service to register for push. This allows registration to be done on app install instead of on-first-run. The webworks portion will receive any pushes.
```
Comment (bryanhiggins) [inline pushHeadless/README.md]: Capitalization of WebWorks is incorrect

File: pushHeadless/README.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+Consult the [PushService](http://developer.blackberry.com/html5/apis/beta/blackberry.push.pushservice.html) and [PushPayload](http://developer.blackberry.com/html5/apis/beta/blackberry.push.pushpayload.html) classes from the API reference for useful examples and a full description of how to use the
+push APIs.
```
Comment (bryanhiggins) [inline pushHeadless/README.md]: Why is there a line break here?

File: pushHeadless/README.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+There is also a valuable developer guide on how to write a push-enabled application using BlackBerry WebWorks that you can find [here](https://developer.blackberry.com/html5/documentation/push_service.html).
+
+The developer guide offers the following topics:
+
+1. An overview of push and the Push Service architecture
+2. The requirements for creating a full push solution
+3. How to download, build, and configure the sample application when it's loaded on your BlackBerry 10 device (This is also described below.)
```
Comment (bryanhiggins) [inline pushHeadless/README.md]: remove comma after build

File: pushHeadless/pushCollectorHeadless/bar-descriptor.xml [hunk 4]
```diff
@@ -0,0 +1,99 @@
+<?xml version="1.0" encoding="utf-8" standalone="no"?>
+<!--
+  
+   Copyright (c) 2011, 2012, 2013 Research In Motion Limited.
```
Comment (bryanhiggins) [inline pushHeadless/pushCollectorHeadless/bar-descriptor.xml]: 2014 BlackBerry Ltd

File: pushHeadless/pushCollectorHeadless/bar-descriptor.xml [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    <!-- Locale support -->
+    <asset path="translations" dest="qm">
+        <include name="*.qm"/>
+    </asset>
+    
+    <!-- Request permission to execute native code.  Required for native applications. -->
+    <permission system="true">run_native</permission>
+    <env var="LD_LIBRARY_PATH" value="app/native/lib:/usr/lib/qt4/lib"/>
+    
+</qnx>
```
Comment (bryanhiggins) [inline pushHeadless/pushCollectorHeadless/bar-descriptor.xml]: is pushCollectorHeadless/icon.png used by WebWorks?

File: pushHeadless/pushCollectorHeadless/readme.txt [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+========================================================================
+Description.
+
+Push Collector Headless is the headless part of the Push Collector
+application. It is responsible for handling all push-related logic,
+including:
+
+- Creating and destroying channels
+- Processing incoming push notifications
```
Comment (bryanhiggins) [inline pushHeadless/pushCollectorHeadless/readme.txt]: Isn't the point of this that WebWorks processes incoming push and the native side only does registration?

File: pushHeadless/pushCollectorHeadless/readme.txt [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+========================================================================
+Requirements:
+
+BlackBerry 10.2 Native SDK
+
+========================================================================
+Running the example:
+
+The headless application cannot be launched directly by the user. To
+launch it, run the Push Collector application. The Push Collector uses
```
Comment (bryanhiggins) [inline pushHeadless/pushCollectorHeadless/readme.txt]: Is this relevant to WebWorks?

File: pushHeadless/pushCollectorHeadless/src/AppHeadless.cpp [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+    , m_shouldRegisterToLaunch(false)
+      , m_shouldUnregisterFromLaunch(false)
+{
+    m_configuration = Configuration::getInstance();
+
+    // By default we initialize the config using settings for a BES push.
+    // If doing a BIS push, comment out the initConfigForBES call in favour of the initConfigForBIS call.
+    // Don't forget to verify your settings in the below two calls.
+    initConfigForBES();
+    //	initConfigForBIS();
```
Comment (bryanhiggins) [inline pushHeadless/pushCollectorHeadless/src/AppHeadless.cpp]: Can this not be done by configuration? This implies that in order to use consumer push, the WebWorks dev needs to download Momentics, copy the project over, import it, uncomment this line and compile.

File: pushHeadless/pushCollectorHeadless/src/AppHeadless.cpp [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+void AppHeadless::sendStatus(const QString &status, const QString &message, int code)
+{
+    //This sample will simply send a status to the console, but this can be changed
+    //to send to the UI via a text file in the share or websockets.
+    qDebug() << status << message << code;
+}
+
+void AppHeadless::initConfigForBES()
+{
+    m_configuration->setProviderApplicationId("myAppID");
```
Comment (bryanhiggins) [inline pushHeadless/pushCollectorHeadless/src/AppHeadless.cpp]: All of these hard coded values should be part of configuration.

File: pushHeadless/www/config.xml [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+            <!-- see https://developer.blackberry.com/native/documentation/cascades/device_platform/headless_apps/create_headless_app.html#modifyentrypoints -->
+            <entryPoint id="HeadlessUI">
+                <entryPointType>Qnx/WebKit</entryPointType>
+            </entryPoint>
+            <entryPoint id="HeadlessService">
+                <name>Headless Service</name>
+                <entryPointType>Qnx/Elf</entryPointType>
+                <permission system="true">_sys_run_headless</permission>
+                <permission>read_device_identifying_information</permission>
+                <permission>access_shared</permission>
```
Comment (bryanhiggins) [inline pushHeadless/www/config.xml]: What does it need access_shared for?

File: pushHeadless/www/config.xml [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+                <invoke-target-type>application</invoke-target-type>
+                <entry-point-id>HeadlessUI</entry-point-id>
+                <filter>
+                    <action>bb.action.PUSH</action>
+                    <mime-type>application/vnd.push</mime-type>
+                </filter>
+            </invoke-target>
+        </config-file>
+    </platform>
+</widget>
```
Comment (bryanhiggins) [inline pushHeadless/www/config.xml:57]: Shouldn't use_consumer_push be somewhere in this file along with instructions to uncomment for enterprise push?

File: pushHeadless/www/index.html [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+<html>
+    <head>
+        <meta charset="utf-8" />
+        <meta name="format-detection" content="telephone=no" />
+        <meta name="viewport" content="user-scalable=no, initial-scale=1, maximum-scale=1, minimum-scale=1, width=device-width, height=device-height, target-densitydpi=device-dpi" />
+        <link rel="stylesheet" type="text/css" href="css/index.css" />
+        <title>Hello World</title>
+    </head>
+    <body>
+        <div class="app">
```
Comment (bryanhiggins) [inline pushHeadless/www/index.html:29]: This could be cleaned up. I don't think its necessary to show the Apache Cordova logo or "WebWorks Ready".

File: pushHeadless/www/js/index.js [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    },
+    // deviceready Event Handler
+    //
+    // The scope of 'this' is the event. In order to call the 'receivedEvent'
+    // function, we must explicity call 'app.receivedEvent(...);'
+    onDeviceReady: function() {
+        var ops = {
+            invokeTargetId : "com.example.pushCollector.invoke.push",
+            appId : "myAppID"
+        };
```
Comment (bryanhiggins) [inline pushHeadless/www/js/index.js:39]: It would be better to read this from configuration.

**Comment by bryanhiggins** [comment]
The native side should be reviewed by Shadid. Sorry, I don't know his GitHub user name.

File: pushHeadless/pushCollectorHeadless/bar-descriptor.xml (hunk 5 above)
Comment (jamesjhedley) [inline pushHeadless/pushCollectorHeadless/bar-descriptor.xml]: This file will end up being removed. It's not needed.

File: pushHeadless/pushCollectorHeadless/readme.txt (hunk 6 above)
Comment (jamesjhedley) [inline pushHeadless/pushCollectorHeadless/readme.txt]: Missed this. Should be removed or updated to reflect our sample.

File: pushHeadless/pushCollectorHeadless/src/AppHeadless.cpp (hunk 8 above)
Comment (jamesjhedley) [inline pushHeadless/pushCollectorHeadless/src/AppHeadless.cpp]: Technically they just need to update the config elements in this file (or a property file) and run make in the pushHeadless directory. Make comes with bbndk i believe. They however, may not have bbndk.  I'll definitely look into possibly using some sort of property file that can be read at runtime to avoid any rebuilding of the native portion.

File: pushHeadless/pushCollectorHeadless/src/AppHeadless.cpp (hunk 9 above)
Comment (jamesjhedley) [inline pushHeadless/pushCollectorHeadless/src/AppHeadless.cpp]: The reason i left it separate was to allow the developer to clearly see what is/isn't required for each type of push. When i was learning push i was VERY confused as to what i actually needed to provide when configuring. For BES all you really need is an AppId, unless you are working with the Push SDK and have an initiator URL and credentials. I just don't want to give them a configuration page with a bunch of empty properties and hope they know what to fill in for each.

I do however agree that we should try and make this as easy for the user as possible. No building, Easy setup.
In order to do that we need to distinguish what fields are required for each type of push and when each optional parameter is needed.

File: pushHeadless/www/config.xml (hunk 10 above)
Comment (jamesjhedley) [inline pushHeadless/www/config.xml]: Shouldnt need this. Good catch

File: pushHeadless/www/config.xml (hunk 11 above)
Comment (jamesjhedley) [inline pushHeadless/www/config.xml:57]: Definitely a good idea.

File: pushHeadless/www/index.html (hunk 12 above)
Comment (jamesjhedley) [inline pushHeadless/www/index.html:29]: Hmm, i'm torn on this one. If they are new to WebWorks i can see this definitely being useful and If they are a WebWorks Vet they are going to simply replace their www with ours anyways. But, yea; None of the other samples keep the logo and webworks ready stuff. Probably is a good idea to keep it as clean as possible and remove.

File: pushHeadless/www/js/index.js (hunk 13 above)
Comment (jamesjhedley) [inline pushHeadless/www/js/index.js:39]: Not sure i fully understand what you mean. Do you mean it would be better to read this from the proposed "configuration property file"? Would be awesome if we could set it up so both the web and native code read from the same file. One stop configuration. No editing several files like all the other samples.

**Comment by jamesjhedley** [comment]
Still need to update WebWorks side to also read from config.

**Comment by jamesjhedley** [comment]
Before merging this needs to have the changes from here ported https://github.com/blackberry/webworks-cli/pull/18/files

**Comment by jamesjhedley** [comment]
BIS push passed for me. Getting steve to test as well.

**Comment by jamesjhedley** [comment]
Rebased, Squashed.
Can someone take a look at the README and see if i missed anything? After that we should be good to merge.

**Comment by jamesjhedley** [comment]
Can somebody please merge this and get it into blackberry repo?
